src/jit/hwintrinsicxarch.cpp

```diff
@@ -295,6 +336,135 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
     return arg;
 }
 
+//------------------------------------------------------------------------
+// immUpperBoundOfHWIntrinsic: get the max imm-value of non-full-range IMM intrinsic
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//
+// Return Value:
+//     the max imm-value of non-full-range IMM intrinsic
+//
+int Compiler::immUpperBoundOfHWIntrinsic(NamedIntrinsic intrinsic)
+{
+    assert(categoryOfHWIntrinsic(intrinsic) == HW_Category_IMM);
+    switch (intrinsic)
+    {
+        case NI_AVX_Compare:
+        case NI_AVX_CompareScalar:
+            return 31; // enum FloatComparisonMode has 32 values
+
+        default:
+            assert((flagsOfHWIntrinsic(intrinsic) & HW_Flag_FullRangeIMM) != 0);
+            return 255;
+    }
+}
+
+//------------------------------------------------------------------------
+// impNonConstFallback: convert certain SSE2/AVX2 shift intrinsic to its semantic alternative when the imm-arg is
+// not a compile-time constant
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//    simdType   -- Vector type
+//    baseType   -- base type of the Vector128/256<T>
+//
+// Return Value:
+//     return the IR of semantic alternative on non-const imm-arg
+//
+GenTree* Compiler::impNonConstFallback(NamedIntrinsic intrinsic, var_types simdType, var_types baseType)
+{
+    assert((flagsOfHWIntrinsic(intrinsic) & HW_Flag_NoJmpTableIMM) != 0);
+    switch (intrinsic)
+    {
+        case NI_SSE2_ShiftLeftLogical:
+        case NI_SSE2_ShiftRightArithmetic:
+        case NI_SSE2_ShiftRightLogical:
+        case NI_AVX2_ShiftLeftLogical:
+        case NI_AVX2_ShiftRightArithmetic:
+        case NI_AVX2_ShiftRightLogical:
+        {
+            GenTree* op2 = impPopStack().val;
+            GenTree* op1 = impSIMDPopStack(simdType);
+            GenTree* tmpOp =
+                gtNewSimdHWIntrinsicNode(TYP_SIMD16, op2, NI_SSE2_ConvertScalarToVector128Int32, TYP_INT, 16);
+            return gtNewSimdHWIntrinsicNode(simdType, op1, tmpOp, intrinsic, baseType, genTypeSize(simdType));
+        }
+
+        default:
+            unreached();
+            return nullptr;
+    }
+}
+
+//------------------------------------------------------------------------
+// isImmHWIntrinsic: check the intrinsic is a imm-intrinsic overload or not
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//    lastOp     -- the last operand of the intrinsic that may point to the imm-arg
+//
+// Return Value:
+//     return if the intrinsic is a imm-intrinsic overload or not because some IMM-intrinsic (HW_Flag_MaybeIMM) have
+//     both imm and vector overloads
+//
+bool Compiler::isImmHWIntrinsic(NamedIntrinsic intrinsic, GenTree* lastOp)
+{
+    if (categoryOfHWIntrinsic(intrinsic) != HW_Category_IMM)
+    {
+        return false;
+    }
+
+    if ((flagsOfHWIntrinsic(intrinsic) & HW_Flag_MaybeIMM) != 0 && genActualType(lastOp->TypeGet()) != TYP_INT)
```

Comment:
It is not obvious why this is correct - it needs comments. This relies on the observation that *all* the `HW_Flag_MaybeIMM` cases have *either* an immmediate integer argument, *or* a non-integer argument, but not a non-constant integer argument.
Perhaps your "Return Value" comment should say:

Return true iff the intrinsics is an imm-intrinsic overload.
Note that some intrinsics, with HW_Flag_MaybeIMM set, have both imm (integer immediate) and vector (i.e. non-TYP_INT) overloads.

Reply:
Thank you! Will do.